Make File json encodable

keith opened the pull request:
This is useful for printing things as json like cquery output.

tetromino reviewed (changes requested):
In principle, agree.

> Currently this encodes the object as the path directly. I'm not sure if that's accurate enough for all cases?

Good question. @comius - should it instead be encoded as `{"root": "...", "short_path": "..."}`? Or denormalized but possibly more useful as `{"root": "...", "path": "..."}`?

----

As for implementation - we already have the `Json.Encodable` interface, I would suggest using it (and make it take an explicit `StarlarkSemantics`) instead of making `Json.java` import starlarkbuildapi.

----

Follow-up: what about label objects?

keith commented:
> As for implementation - we already have the Json.Encodable interface

ah nice thanks I missed this. I switched to that here, although I made it pass through StarlarkSemantics too, since it seems like that's something we need in this case and probably others. Are there internal google implementors of this interface? since there aren't any public ones today.

I also noticed that we might want the `appendQuoted` helper from the implementers of this interface. I skipped it here but I think that might be required if there are non-ascii characters in the file path?

> Follow-up: what about label objects?

probably would be nice to have too. I can implement that if this approach looks good

keith commented:
I updated this to include `path`, `short_path`, and `root`. I also updated the Encodable interface so that we shared the logic for how to escape strings before encoding them, happy to address any feedback here

keith commented:
@tetromino friendly ping
